compiler/damlc/lib/DA/Cli/Damlc.hs

```diff
@@ -851,40 +902,387 @@ installDepsAndInitPackageDb opts (InitPkgDb shouldInit) =
+  { bmSdkVersion :: PackageSdkVersion
+  , bmName :: LF.PackageName
+  , bmVersion :: LF.PackageVersion
+  , bmDataDeps :: [FilePath]
+      -- ^ not canonicalized
+  , bmSourceDaml :: FilePath
+      -- ^ not canonicalized
+  , bmOutput :: Maybe FilePath
+      -- ^ not canonicalized
+  }
+
+-- Version of getDamlFiles from Dar.hs that assumes source is a folder, and runs in IO rather than action
+-- Throws an error if its a path to a file
+getDamlFilesBuildMulti :: FilePath -> IO (Map.Map FilePath BSL.ByteString)
+getDamlFilesBuildMulti path = do
+  exists <- doesDirectoryExist path
+  -- TODO[SW]: find a way to fix this
+  -- It isn't fully possible, since we need to be able to parse the files to check their imports, and we can't parse other sdk versions.
+  -- The best we can probably do is find all daml files in the daml.yaml directory?
+  unless exists $ error "multi-package build does not support daml file source paths. Please specify a directory containing your daml files instead."
+  files <- fmap fromNormalizedFilePath <$> damlFilesInDir path
+  Map.fromList <$> traverse (\file -> (makeRelative path file,) <$> BSL.readFile file) files
+
+-- Extract the name/version string and package ID of a given dalf in a dar.
+entryToDalfData :: ZipArchive.Entry -> Maybe (String, LF.PackageId)
+entryToDalfData entry = do
+  let fileName = takeBaseName $ ZipArchive.eRelativePath entry
+  lastDash <- listToMaybe $ reverse $ elemIndices '-' fileName
+  pure $ second (LF.PackageId . T.pack . tail) $ splitAt lastDash fileName
+
+dalfDataKey :: LF.PackageName -> LF.PackageVersion -> String
+dalfDataKey name version = T.unpack (LF.unPackageName name <> "-" <> LF.unPackageVersion version)
+
+-- Given an archive, returns the packagename-packageversion string with package ID for all dalfs, and the file path and file contents for all daml files.
+readDarStalenessData :: ZipArchive.Archive -> (Map.Map String LF.PackageId, Map.Map FilePath BSL.ByteString)
+readDarStalenessData archive =
+  bimap Map.fromList Map.fromList . partitionEithers $
+    flip mapMaybe (ZipArchive.zEntries archive) $ \entry ->
+      case takeExtension $ ZipArchive.eRelativePath entry of
+        ".dalf" -> Left <$> entryToDalfData entry
+        ".daml" ->
+          -- Drop the first directory in the path, as thats the folder containing the entire package contents.
+          -- Previously structured this way to support multiple packages in one dar.
+          let relPackagePath = joinPath $ tail $ splitPath $ ZipArchive.eRelativePath entry
+           in Just $ Right (relPackagePath, ZipArchive.fromEntry entry)
+        _ -> Nothing
+
+-- | Gets the package id of a Dar in the given project ONLY if it is not stale.
+getValidPackageId :: IDELogger.Logger -> FilePath -> BuildMultiPackageConfig -> [(LF.PackageName, LF.PackageVersion, LF.PackageId)] -> IO (Maybe LF.PackageId)
+getValidPackageId logger path BuildMultiPackageConfig {..} sourceDepPids = do
```

Comment:
Maybe a better name than "valid" would be good for readability - "getPackageIdIfFresh"? I didn't realize how much work this was doing to fulfill the staleness detection on my initial read through.

Reply:
Fair point, i like that name